include/pybind11/numpy.h

```diff
@@ -1436,6 +1436,24 @@ struct npy_format_descriptor<T, enable_if_t<is_same_ignoring_cvref<T, PyObject *
     static pybind11::dtype dtype() { return pybind11::dtype(/*typenum*/ value); }
 };
 
+template <>
```

Comment:
Could you please try if we can avoid the code duplication?

Idea 1:

Work on the `enable_if_t` in line 1431, something like (untested) `|| std::is_same<T, handle>::value || std::is_same<T, object>::value` (but also include your `sizeof` condition). If you feel ambitious, could `std::is_base_of<handle, T>::value` work? I'm totally not sure myself.

Idea 2:

Use inheritance.

Reply:
Done. I was hesitant to use `std::is_base_of<handle, T>::value` though because that would also capture `py::str`, `py::list` etc which doesn't sound like a good idea. It would allow constructing e.g. a `py::array_t<py::str>` which does not actually check whether the contained objects are in fact strings.